### PR TITLE
Add zipdeploy tasks for publishing from VS

### PR DESCRIPTION
**Pull request by ankitkumarr**
Most of the changes come from -- https://github.com/Azure/azure-functions-vs-build-sdk/tree/v3.x/src/Microsoft.NET.Sdk.Functions.MSBuild

Fixes https://github.com/Azure/azure-functions-dotnet-worker/issues/413

### CODE REVIEW

File: sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets
Comment (ankitkumarr) [inline sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.targets:261]: Removing this property "DeeployOnBuild" so that ZipDeploy and other tasks don't get invoked from the "WorkerExtensions" project.

File: sdk/Sdk/Tasks/ZipDeploy/CreateZipFileTask.cs
Comment (ankitkumarr) [inline sdk/Sdk/Tasks/ZipDeploy/CreateZipFileTask.cs:14]: I kept the namespaces same as what they are in https://github.com/Azure/azure-functions-vs-build-sdk/tree/main/src/Microsoft.NET.Sdk.Functions.MSBuild

I went back and forth on whether to update these to be "Microsoft.Azure.Functions.Worker.Sdk.MSBuild" or not. I thought maybe leave it to indicate that this is a copy (with some changes to get rid of warning like nullables and not using Resources.resx).

I am open to counter arguments here. Really not a big deal either way, as this isn't some exposed API.

File: sdk/Sdk/Tasks/ZipDeploy/StringMessages.cs
Comment (ankitkumarr) [inline sdk/Sdk/Tasks/ZipDeploy/StringMessages.cs:6]: vs-build-sdk uses Resources.resx for messages -- https://github.com/Azure/azure-functions-vs-build-sdk/blob/main/src/Microsoft.NET.Sdk.Functions.MSBuild/Properties/Resources.resx

But I just put them here for simplicity.

File: sdk/Sdk/Tasks/ZipDeploy/CreateZipFileTask.cs
Comment (ankitkumarr) [inline sdk/Sdk/Tasks/ZipDeploy/CreateZipFileTask.cs:12]: Added this header in all the files that came from vs-build-sdk.

File: sdk/Sdk/Tasks/ZipDeploy/ZipDeployTask.cs
Comment (SeanFeldman) [inline sdk/Sdk/Tasks/ZipDeploy/ZipDeployTask.cs:57]: I recall it's better to replace those two with the following:

```suggestion
                return t.GetAwaiter().GetResult();
```

File: sdk/Sdk/Tasks/ZipDeploy/ZipDeployTask.cs
Comment (SeanFeldman) [inline sdk/Sdk/Tasks/ZipDeploy/ZipDeployTask.cs:71]: This looks really funky. This is where the new syntax is so nice:

```suggestion
                if (publishUrl!.EndsWith("/") is false)
```

File: sdk/Sdk/Tasks/ZipDeploy/ZipDeployTask.cs
Comment (SeanFeldman) [inline sdk/Sdk/Tasks/ZipDeploy/ZipDeployTask.cs:130]:
```suggestion
                        Log.LogError(string.Format(StringMessages.ZipDeployFailureErrorMessage, zipDeployPublishUrl, status));
```

File: sdk/Sdk/Tasks/ZipDeploy/ZipDeploymentStatus.cs
Comment (SeanFeldman) [inline sdk/Sdk/Tasks/ZipDeploy/ZipDeploymentStatus.cs:59]:
```suggestion
                        _log.LogMessage(string.Format(StringMessages.DeploymentStatus, Enum.GetName(typeof(DeployStatus), deployStatus)));
```

File: sdk/Sdk/Tasks/ZipDeploy/ZipDeployTask.cs
Comment (SeanFeldman) [inline sdk/Sdk/Tasks/ZipDeploy/ZipDeployTask.cs:94]:
```suggestion
                Log.LogMessage(MessageImportance.High, string.Format(StringMessages.PublishingZipViaZipDeploy, zipToPublishPath, zipDeployPublishUrl));
```

File: sdk/Sdk/Tasks/ZipDeploy/ZipDeployTask.cs
Comment (SeanFeldman) [inline sdk/Sdk/Tasks/ZipDeploy/ZipDeployTask.cs:106]:
```suggestion
                    Log.LogError(string.Format(StringMessages.ZipDeployFailureErrorMessage, zipDeployPublishUrl, response.StatusCode));
```

File: sdk/Sdk/Tasks/ZipDeploy/Http/HttpResponseMessageForStatusCode.cs
Comment (SeanFeldman) [inline sdk/Sdk/Tasks/ZipDeploy/Http/HttpResponseMessageForStatusCode.cs:31]:
```suggestion
            return Enumerable.Empty<string>();
```

File: sdk/Sdk/Tasks/ZipDeploy/ZipDeploymentStatus.cs
Comment (SeanFeldman) [inline sdk/Sdk/Tasks/ZipDeploy/ZipDeploymentStatus.cs:124]: Shouldn't this re-throw?

```suggestion
                        throw;
```

**Review by SeanFeldman — COMMENTED** [review]
A few minor things

**Comment by ankitkumarr** [comment]
@SeanFeldman thank you so much for the review :). I agree with all your changes.
But we are trying to not change the logic here unless required to build or unblock this scenario. Ideally all of your changes would be directly in https://github.com/Azure/azure-functions-vs-build-sdk/tree/v3.x/src/Microsoft.NET.Sdk.Functions.MSBuild/Tasks

I have opened an item to consolidate the logic and added a note --
https://github.com/Azure/azure-functions-dotnet-worker/issues/419

You are more than welcome to open a PR in the "azure-functions-vs-sdk" repository for the changes, so we can accept there and then port it over. If not, we can track the changes as part of the workitem opened.

File: sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.Publish.targets
Comment (ankitkumarr) [inline sdk/Sdk/Targets/Microsoft.Azure.Functions.Worker.Sdk.Publish.targets:76]: Actually I think this block and the tasks referenced are already there in "Microsoft.Azure.Functions.Worker.Sdk.targets" so I will remove these.

File: sdk/Sdk/Tasks/ZipDeploy/Http/HttpResponseMessageWrapper.cs
Comment (phil-allen-msft) [inline sdk/Sdk/Tasks/ZipDeploy/Http/HttpResponseMessageWrapper.cs:17]: Should be IDisposable, since HttpResponseMessage is IDisposable (I realize this issue also exists in the other repo)

File: sdk/Sdk/Tasks/ZipDeploy/ZipDeploymentStatus.cs
Comment (phil-allen-msft) [inline sdk/Sdk/Tasks/ZipDeploy/ZipDeploymentStatus.cs:53]: For me, at least, && implies a new line once the total expression is 'long'.  Would that help readability?

File: sdk/Sdk/Tasks/ZipDeploy/CreateZipFileTask.cs
Comment (fabiocav) [reply sdk/Sdk/Tasks/ZipDeploy/CreateZipFileTask.cs:14]: I'm good with this. The less we diverge, the better at this point.

File: sdk/Sdk/Tasks/ZipDeploy/Http/HttpResponseMessageForStatusCode.cs
Comment (fabiocav) [reply sdk/Sdk/Tasks/ZipDeploy/Http/HttpResponseMessageForStatusCode.cs:31]: @ankitkumarr we should take this feedback into the original files.

File: sdk/Sdk/Tasks/ZipDeploy/Http/HttpResponseMessageWrapper.cs
Comment (fabiocav) [reply sdk/Sdk/Tasks/ZipDeploy/Http/HttpResponseMessageWrapper.cs:17]: Agreed.

File: sdk/Sdk/Tasks/ZipDeploy/ZipDeployTask.cs
Comment (fabiocav) [inline sdk/Sdk/Tasks/ZipDeploy/ZipDeployTask.cs:28]: Again, I know this is a problem with the original implementation, but do we want this to report `functions-core-tools`? That's not really accurate.

File: sdk/Sdk/Tasks/ZipDeploy/ZipDeployTask.cs
Comment (fabiocav) [reply sdk/Sdk/Tasks/ZipDeploy/ZipDeployTask.cs:57]: This would indeed ensure cleaner exceptions.

File: sdk/Sdk/Tasks/ZipDeploy/ZipDeploymentStatus.cs
Comment (fabiocav) [reply sdk/Sdk/Tasks/ZipDeploy/ZipDeploymentStatus.cs:53]: +1

File: sdk/Sdk/Tasks/ZipDeploy/ZipDeploymentStatus.cs
Comment (fabiocav) [reply sdk/Sdk/Tasks/ZipDeploy/ZipDeploymentStatus.cs:134]: Yes.

**Review by fabiocav — APPROVED** [review]
I'm good with the changes going in as is, but it would be great to collect the feedback provided on the files here for enhancements in the original implementation, we can then pull the changes back here.

**Comment by ankitkumarr** [comment]
@fabiocav, yes I agree. I have created a follow-up item. Once this is checked in and we unblock the scenario, I will follow-up with a review of the target with have in the "vs-sdk" repository and incorporate all the suggestions made here, and then port them over here.

For now, I am merging this as is.

**Pull request merged**
